### PR TITLE
Optim foreach cleanup for Adagrad

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* #70229
* #69982
* __->__ #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to Adagrad optimizer + cleanup

Differential Revision: [D33767863](https://our.internmc.facebook.com/intern/diff/D33767863)

### CODE REVIEW

File: torch/optim/adagrad.py [hunk 1]
```diff
@@ -103,13 +112,135 @@ def step(self, closure=None):
...
+    if grad_indices.numel() == 0 or values.numel() == 0:
+        return torch.empty_like(grad)
+    return torch.sparse_coo_tensor(grad_indices, values, size)
+
+
+def _single_tensor_adagrad(params: List[Tensor],
+                           grads: List[Tensor],
+                           state_sums: List[Tensor],
+                           state_steps: List[int],
+                           has_sparse_grad: bool = False,
```
Comment (mikaylagawarecki) [inline torch/optim/adagrad.py:171]: added an unused kwarg with a default here `has_sparse_grad` so both functions take the same inputs and the above logic looks cleaner, is this ok?

File: torch/optim/adagrad.py [hunk 2]
```diff
@@ -103,13 +116,149 @@ def step(self, closure=None):
...
+                          lr: float,
+                          weight_decay: float,
+                          lr_decay: float,
+                          eps: float):
+
+    # Foreach functions will throw errors if given empty lists
+    if len(params) == 0:
+        return
+
+    if has_sparse_grad:
```
Comment (albanD) [inline torch/optim/adagrad.py:237]: This should never happen given that you already check that in the main function right?

File: torch/optim/adagrad.py [hunk 3]
```diff
@@ -103,13 +116,149 @@ def step(self, closure=None):
...
         return loss
+
+
+def adagrad(params: List[Tensor],
+            grads: List[Tensor],
+            state_sums: List[Tensor],
+            state_steps: List[int],
+            # kwonly args with defaults are not supported by functions compiled with torchscript issue #70627
+            # setting these as kwargs for now as functional API is compiled by torch/distributed/optim
+            has_sparse_grad: bool = False,
```
Comment (albanD) [inline torch/optim/adagrad.py:139]: Why is the default False here? (and not None)
Don't we want to check it if the user didn't provide it? Or raise an error if the user asked for foreach explicitly but didn't provide this?

**Review by albanD — COMMENTED** [review]
Small question about has_sparse_grad. Sounds good otherwise

File: torch/optim/adagrad.py [hunk 4]
```diff
@@ -92,21 +102,157 @@ def step(self, closure=None):
...
+        raise RuntimeError("API has changed, `state_steps` argument must contain a list of singleton tensors")
+
+    if foreach is None:
+        # Placeholder for more complex foreach logic to be added when value is not set
+        foreach = False
+
+    if foreach and torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if has_sparse_grad is None:
```
Comment (albanD) [inline torch/optim/adagrad.py:158]: We don't need to do that if we call into the non-multi-tensor impl. Given that this is not "free" we should skip that.
I think by moving this logic in the multi-tensor code is a good way to do that.

**Review by albanD — COMMENTED** [review]
Looks good. We should just avoid some extra computation that is not needed but good to go otherwise.

**Review by albanD — APPROVED** [review]
Thanks for the update

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767863).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767863).

**Pull request closed without merging**